### ISSUE TITLE
DPDK: An obvious bug in the nfp_net.c: found by gcc-8

### ISSUE THREAD

**Issue by vladzcloudius** [scylladb/seastar#501]
HEAD: 57128167aa94efa3bd62b903d4830cc02f736905
Side effect: compilation with gcc/g++-8 with `--enable-dpdk` is broken

```
CC nfp_net.o
distcc[4859] ERROR: compile /home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c on 127.0.0.7:12006/8,lzo failed
distcc[4859] (dcc_build_somewhere) Warning: remote compilation of '/home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c' failed, retrying locally
distcc[4859] Warning: failed to distribute /home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c to 127.0.0.7:12006/8,lzo, running locally instead
distcc[4859] ERROR: compile /home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c on localhost failed
In file included from /usr/include/string.h:494,
                 from /home/vladz/work/scylla/seastar/build/dpdk/include/rte_memcpy.h:45,
                 from /home/vladz/work/scylla/seastar/build/dpdk/include/rte_ether.h:50,
                 from /home/vladz/work/scylla/seastar/build/dpdk/include/rte_ethdev.h:185,
                 from /home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c:48:
In function ‘memcpy’,
    inlined from ‘nfp_net_read_mac’ at /home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c:602:2,
    inlined from ‘nfp_net_init’ at /home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c:2542:2:
/usr/include/x86_64-linux-gnu/bits/string_fortified.h:34:10: error: ‘__builtin_memcpy’ forming offset [5, 6] is out of the bounds [0, 4] of object ‘tmp’ with type ‘uint32_t’ {aka ‘unsigned int’} [-Werror=array-bounds]
   return __builtin___memcpy_chk (__dest, __src, __len, __bos0 (__dest));
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
/home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c: In function ‘nfp_net_init’:
/home/vladz/work/scylla/seastar/dpdk/drivers/net/nfp/nfp_net.c:599:11: note: ‘tmp’ declared here
  uint32_t tmp;
           ^~~
cc1: all warnings being treated as errors
/home/vladz/work/scylla/seastar/dpdk/mk/internal/rte.compile-pre.mk:138: recipe for target 'nfp_net.o' failed
make[5]: *** [nfp_net.o] Error 1
/home/vladz/work/scylla/seastar/dpdk/mk/rte.subdir.mk:63: recipe for target 'nfp' failed
make[4]: *** [nfp] Error 2
/home/vladz/work/scylla/seastar/dpdk/mk/rte.subdir.mk:63: recipe for target 'net' failed
make[3]: *** [net] Error 2
```

The root cause is this function:
```
static void nfp_net_read_mac(struct nfp_net_hw *hw)
{
	uint32_t tmp;

	tmp = rte_be_to_cpu_32(nn_cfg_readl(hw, NFP_NET_CFG_MACADDR));
	memcpy(&hw->mac_addr[0], &tmp, sizeof(struct ether_addr));

	tmp = rte_be_to_cpu_32(nn_cfg_readl(hw, NFP_NET_CFG_MACADDR + 4));
	memcpy(&hw->mac_addr[4], &tmp, 2);
}
```

`struct ether_addr` is defined in `rte_ether.h`:
```
#define ETHER_ADDR_LEN  6 /**< Length of Ethernet address. */
...
struct ether_addr {
	uint8_t addr_bytes[ETHER_ADDR_LEN]; /**< Addr bytes in tx order */
} __attribute__((__packed__));
```

Therefore the 
```
memcpy(&hw->mac_addr[0], &tmp, sizeof(struct ether_addr));
``` 
is definitelly going to read outside the `tmp`.

**Comment by avikivity** [comment]
Was fixed upstream in 

```
commit 1bcb5ecb179cbca81ff3fe15432326c1c3ca9e9b
Author: Andy Green <andy@warmcat.com>
Date:   Mon May 14 13:00:32 2018 +0800

    net/nfp: fix memcpy out of source range
    
    drivers/net/nfp/nfp_net.c:669:2: error:
    ‘memcpy’ forming offset [5, 6] is out of the bounds [0, 4]
    of object ‘tmp’ with type ‘uint32_t’ {aka ‘unsigned int’}
    [-Werror=array-bounds]
    memcpy(&hw->mac_addr[0], &tmp, sizeof(struct ether_addr));
    
    Fixes: e6decee38209 ("net/nfp: use random MAC address if not configured")
    Cc: stable@dpdk.org
    
    Signed-off-by: Andy Green <andy@warmcat.com>
    Acked-by: Alejandro Lucero <alejandro.lucero@netronome.com>
    Tested-by: Alejandro Lucero <alejandro.lucero@netronome.com>
```

**Comment by avikivity** [comment]
Fixed by 0351626ec6cbc2aa02c34f26c1bc759e9aac4178
